SFS_general/delayline.m

```diff
@@ -59,88 +57,160 @@
 %*****************************************************************************
 
 
-%% ===== Configuration ==================================================
-usefracdelay = conf.usefracdelay;
+%% ===== Configuration ===================================================
+% Check for old configuration
+if isfield(conf, 'usefracdelay')
+    error(['%s: conf.usefracdelay is deprecated, please use conf.delayline', ...
+           ' instead. See SFS_config for details.'],upper(mfilename));
+end
+delay = conf.delayline;
 
 
-%% ===== Computation =====================================================
-% Check if the impulse response is given in SOFA conventions [M C N], or in
-% usual [N C] convention, where
+%% ===== Preparation =====================================================
+% --- Reshape signals ---
+% Check if the signal is an impulse response given in SOFA conventions [M C N],
+% or in usual [N C] convention, where
 % M ... number of measurements
 % C ... number of channels
 % N ... number of samples
 if ndims(sig)==3
-    [M C samples] = size(sig);
+    [M,C,samples] = size(sig);
     channels = M * C;
     % Reshape [M C N] => [N C*M], this will be redone at the end of the function
     sig = reshape(sig,[channels,samples])';
     reshaped = true;
 else
     % Assume standard format [N C]
-    [samples channels] = size(sig);
+    [samples,channels] = size(sig);
     reshaped = false;
 end
+% --- Expand dt and weight ---
 % If only single valued time delay and weight is given, create vectors
 if channels>1 && length(dt)==1, dt=repmat(dt,[1 channels]); end
 if channels>1 && length(weight)==1, weight=repmat(weight,[1 channels]); end
 
-if usefracdelay
-
-    % Additional configuration
-    fracdelay_method = conf.fracdelay_method;
-    rfactor = 100; % resample factor (1/stepsize of fractional delays)
-    Lls = 30;      % length of least-squares factional delay filter
 
-    % Defining a temporary conf struct for recursive calling of delayline
-    conf2.usefracdelay = false;
+%% ===== Resampling ======================================================
+% The resampling is applied independently from the actual fractional/integer
+% delay handling performed in the next step. The resampling is redone at the end
+% of the file.
+% If resampling is used together with the integer delay filter, this is already
+% a usage of fractional delay due to the upsampling.
+%
+switch delay.resampling
+    case 'none'
+        rfactor = 1.0;
+    case 'matlab'
+        rfactor = delay.resamplingfactor;
+        sig = resample(sig,rfactor,1);
+    case 'pm'
+        % === Parks-McClellan linear phase FIR filter ===
+        rfactor = delay.resamplingfactor;
+        a = [1 1 0 0];
+        f = [0.0 0.9/rfactor 1/rfactor 1.0];
+        b = firpm(delay.resamplingorder,f,a);
```

Comment:
I don't remember how it got there, but having a fixed order for the filter design independent from the used resampling order is definitely a bad idea. I has to scale with the resampling factor.